Check for Tekton installed before running the manager

alicerum opened the pull request:
# Changes

<!-- 🎉🎉🎉 Thank you for the PR!!! 🎉🎉🎉 -->

<!-- Describe your changes here- ideally you can get that description straight from
your descriptive commit message(s)! -->

Right now reconciler tries to watch TaskRun objects when manager starts. If tekton is not installed, controller crashloops.
This change brings check for TaskRuns existing in the earlier stage and exits with explicit error and explanation.
This way user should know perfectly well what the problem is and how to fix it.

We don't want to let reconcilers work if tekton is not installed, because this behavior might confuse the user and make it more difficult to find out what's wrong. In this situation we want to exit asap and have a very clear explanation of the reason of the failure.

I am keeping this check in `main.go`, because I am not sure of any place that is better suited for it, I figured it might be the best to run it as soon as `manager` object is created. If anybody thinks there is a better suitable place for it, let me know, please.

<!-- If this PR fixes a GitHub issue, please mention it like so: -->

Fixes #734

# Submitter Checklist

- [ ] Includes tests if functionality changed/was added
- [x] Includes docs if changes are user-facing
- [x] [Set a kind label on this PR](https://prow.k8s.io/command-help#kind)  
- [x] Release notes block has been filled in, or marked NONE

See [the contributor guide](https://github.com/shipwright-io/build/blob/main/CONTRIBUTING.md)
for details on coding conventions, github and prow interactions, and the code review process.

# Release Notes

<!--
Describe any user facing changes here, or delete this block.

Examples of user facing changes:
- API changes
- Bug fixes
- Any changes in behavior
- Changes requiring upgrade notices or deprecation warnings

For pull requests with a release note:

```release-note
Your release note here
```

For pull requests that require additional action from users switching to the new release, include the string "action required" (case insensitive) in the release note:

```release-note
action required: your release note here
```

For pull requests that don't need to be mentioned at release time, use the `/release-note-none` Prow command to add the `release-note-none` label to the PR. You can also write the string "NONE" as a release note in your PR description:

```release-note
NONE
```
-->
```release-note
Controller now checks for `TERMINATION_LOG_PATH` environmental variable in order to put reason of termination into it. Default value is `/dev/termination-log`.
```

alicerum commented:
small proposal: generate event about this failure for the user

SaschaSchwarze0 commented on pkg/config/config.go:
```diff
@@ -59,6 +59,9 @@ const (
 	kubeAPIBurst = "KUBE_API_BURST"
 	kubeAPIQPS   = "KUBE_API_QPS"
+
+	terminationLogPathDefault = "/dev/termination-log"
```
Does that work and make sense = writing to the `/dev` directory related to devices ?

alicerum replied:
this is the kubernetes recommended way and default termination log location. can be overriden via deployment

SaschaSchwarze0 replied:
Indeed, sry, was not aware of a default. I though one always must specify `terminationMessagePath`. :-)

SaschaSchwarze0 reviewed (commented):
Should we specify the terminate message path for our controller container in https://github.com/shipwright-io/build/blob/main/deploy/500-controller.yaml ?

alicerum commented:
So, we cannot create events in the pod, because event has to have a context of an object reference. This is why events are recommended to be fired from the reconciler. Since this event would be about a pod not starting, we have no good way of triggering it.

But, I added the logic to write mistake to the termination log and added it's path to the configuration. It can be configured via the `TERMINATION_LOG_PATH` environment variable, and it's default value is `/dev/termination log`.
This way, error is getting into the status of a failed controller:

```
    State:          Waiting
      Reason:       CrashLoopBackOff
    Last State:     Terminated
      Reason:       Error
      Message:      Cannot start manager: Tekton Pipelines are not installed on the cluster
      Exit Code:    1
      Started:      Wed, 02 Jun 2021 14:37:21 +0200
      Finished:     Wed, 02 Jun 2021 14:37:28 +0200

```

alicerum commented:
> Should we specify the terminate message path for our controller container in https://github.com/shipwright-io/build/blob/main/deploy/500-controller.yaml ?

It's default value is `/dev/termination-log`, we should specify in case we want to use other location though. For now, I'm just using defaults.

SaschaSchwarze0 reviewed (commented):
Now that I know about the default `/dev/termination-log`. :-) Is there a use case at all where this would be customized? If so, we should include it in our [documentation](https://github.com/shipwright-io/build/blob/main/docs/configuration.md).

alicerum commented:
If there will be any necessity for this, I do not know. But, this path can be configured via deployment, and due to that, i think, it should also be configurable for the controller application itself. Maybe something well change one day, and it will work differently.

I will update documentation, thanks!

otaviof reviewed (commented):
Good move!

alicerum commented:
/assign @imjasonh

imjasonh commented on cmd/shipwright-build-controller/main.go:
```diff
@@ -113,3 +136,28 @@ func main() {
+// into the termination log file, making it a clear explanation for
+// pod failure
+func writeTerminationMessage(terminationLogPath string, msg string) error {
+	return ioutil.WriteFile(terminationLogPath, []byte(msg), 0644)
```
I'm not sure this method adds much, compared to just calling WriteFile with a comment explaining what it does

alicerum replied:
i fixed that!

imjasonh reviewed (commented):
/lgtm

SaschaSchwarze0 reviewed (approved):
/approve

otaviof commented:
/lgtm
